shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,773 @@
+  const destinationPermissions = {};
+  {
+    const clause = new Clause();
+    destinationPermissions[clause.hash()] = new MembranedPermissionSet(permissionSet, clause);
   }
 
-  var openerIdentityId = edges.terminalEdge.sharer;
-  var edgesByRecipient = edges.edgesByRecipient;
-  var owner = "OwningAccount";
+  const permissionsMap = {};
+  // Map<vertexId, Map<membraneHash, MembranedPermissionSet>>.
+  // May from vertex ID to permissions that we've already shown flow from that vertex to our
+  // destination vertex.
 
-  var permissionsMap = {};
-  // Keeps track of the permissions that the opener receives from each user. The final result of
-  // our computation will be stored in permissionsMap[owner].
+  const vertexStack = []; // Vertex IDs that we need to explore.
 
-  var userStack = [openerIdentityId];
+  permissionsMap[destinationId] = destinationPermissions;
+  vertexStack.push(destinationId);
 
-  var openerAttenuation = roleAssignmentPermissions({allAccess: null}, viewInfo);
-  edges.terminalEdge.roleAssignments.forEach(function(roleAssignment) {
-    openerAttenuation.intersect(roleAssignmentPermissions(roleAssignment, viewInfo));
-  });
-  permissionsMap[openerIdentityId] = openerAttenuation;
-
-  while (userStack.length > 0) {
-    var recipient = userStack.pop();
-    if (edgesByRecipient[recipient]) {
-      edgesByRecipient[recipient].forEach(function (inEdge) {
-        var sharer = inEdge.sharer;
-        var needToPush = false;
-        if (!permissionsMap[sharer]) {
-          permissionsMap[sharer] = new PermissionSet();
-          needToPush = true;
-        }
+  while (vertexStack.length > 0) {
+    const vertexId = vertexStack.pop();
+    let incomingEdges = [];
 
-        var newPermissions = roleAssignmentPermissions({allAccess: null}, viewInfo);
-        inEdge.roleAssignments.forEach(function(roleAssignment) {
-          newPermissions.intersect(roleAssignmentPermissions(roleAssignment, viewInfo));
-        });
-        newPermissions.intersect(permissionsMap[recipient]);
+    function tokenToEdge(token) {
+      let sharerId = "i:" + token.identityId;
+      if (token.parentToken) {
+        sharerId = "t:" + token.parentToken;
+      }
 
-        // Optimization: we don't care about permissions that we've already proven the opener has.
-        if (permissionsMap[owner]) {
-          newPermissions.remove(permissionsMap[owner]);
-        }
+      return {
+        sharerId: sharerId,
+        membranedPermissions: MembranedPermissionSet.fromToken(token, viewInfo),
+      };
+    }
+
+    if (vertexId.slice(0, 2) == "o:") {
+      // Owner. We don't need to do anything.
+      incomingEdges = [];
+    } else if (vertexId.slice(0, 2) == "t:") {
+      const token = context.tokensById[vertexId.slice(2)];
+      if (token) {
+        incomingEdges = [tokenToEdge(token)];
+      }
+    } else if (ownerIdentityIds.indexOf(vertexId.slice(2)) >= 0) {
```

Comment:
I think you should explicitly check here that the vertexId starts with "i:", in case a different kind of vertexId is added later.